### PR TITLE
Add FractionSum Calculation

### LINKED ISSUE #32: Add new `generate.functions.FractionalSum`
Related to PR #31. Maybe needed to progress on Labor projections.

_This issue is continuing a post by @jrising in https://github.com/ClimateImpactLab/open-estimate/pull/31#issuecomment-621043424:_

> While I have no problem with us having a Product calculation, I think there are still reasons to prefer a `FractionalSum` calculation object: you avoid the 1 - X problem, you can check that the fraction has no units and stays within bounds, the formatted code can be cleaner and the argument types more informative. But the partial derivative of a FractionalSum is a Sum of a FractionalSum and a Product, so you would still need to have this Product class anyway.

The goal here is to talk about whether it's needed, what's needed to make it work and what this thing might look like in the interface.

### PR DESCRIPTION
**Pull request by brews**
This PR:
- Adds `FractionSum` calculation.
- Fixes a minor bug in `Production` exception message in `enable_deltamethod()`.

This PR includes docstrings and unit tests for the new feature.

Close #32.

### CODE REVIEW

**Comment by brews** [comment]
Here is the simplest implementation I could think of. Feel free to rip and tear. It's ready for you to take a look at, @jrising .

One question, off the bat: I take in `frac` and `subcalcs` as initializing args. This is the first time I've had to split initializing subcalculations across two arguments. I think I have the `.describe()` right but my `.apply()` implementation feels hackish...?

File: openest/generate/functions.py [hunk 1]
```diff
@@ -443,6 +443,148 @@ def describe():
...
+    frac : ``openest.generate.calculation.Calculation``
+        Result gives weight of first subcalculation. 1 - `frac` is weight of
+        second subcalculation. Must be ≤ 1. The
+    subcalcs : Sequence of ``openest.generate.calculation.Calculation``
+        Calculations to weight. Must be ``len(subcalcs) == 2``. The units of
+        the subcalcs elements must be the same.
+    unshift : bool, optional
+    """
+
+    def __init__(self, frac, subcalcs, unshift=True):
```
Comment (jrising) [inline openest/generate/functions.py:463]: The automatic argument matcher will get confused by splitting out the first of three subcalcs, and there's a second reason to keep them all in one list: you can generalize this to N subcalcs, by saying that they go in the following order:
`fraction1, subcalc1, [fraction2, subcalc2, ...], subcalcN`
where the fraction applied to `subcalcN` is just (1 - (fraction1 + fraction2 + ...)).

If you don't plan on setting this up in the general way, then I think you should just have three separate arguments:
`fraccalc`, `truecalc`, `falsecalc`
or similarly named.

File: openest/generate/functions.py [hunk 2]
```diff
@@ -443,6 +443,148 @@ def describe():
...
+        if len(subcalcs) != 2:
+            raise ValueError('Length of `subcalcs` must be 2')
+
+        fullunitses = subcalcs[0].unitses[:]
+        for ii in range(1, len(subcalcs)):
+            assert subcalcs[0].unitses[0] == subcalcs[ii].unitses[0], "%s <> %s" % (subcalcs[0].unitses[0], subcalcs[ii].unitses[0])
+            fullunitses.extend(subcalcs[ii].unitses)
```
Comment (jrising) [inline openest/generate/functions.py:470]: Also check that frac is unitless.

File: openest/generate/functions.py [hunk 3]
```diff
@@ -443,6 +443,148 @@ def describe():
...
+        mains = []
+        elements = {}
+        alldeps = set()
+        for subcalc in self.subcalcs:
+            elements.update(subcalc.format(lang, *args, **kwargs))
+            mains.append(elements['main'])
+            alldeps.update(elements['main'].dependencies)
+
+        if lang in ['latex', 'julia']:
+            elements['main'] = FormatElement(' * '.join([main.repstr for main in mains]), list(alldeps))
```
Comment (jrising) [inline openest/generate/functions.py:492]: Fix this.

File: openest/generate/functions.py [hunk 4]
```diff
@@ -443,6 +443,148 @@ def describe():
...
+
+        formatting.add_label('product', elements)
```
Comment (jrising) [inline openest/generate/functions.py:494]: Also here.

File: openest/generate/functions.py [hunk 5]
```diff
@@ -443,6 +443,148 @@ def describe():
...
+
+                ``"title"``
+                    Long-length title of this calculation
+
+                ``"description"``
+                    Long-form description of this calculation
+        """
+        infoses = [subcalc.column_info() for subcalc in self.subcalcs]
+        title = 'FractionSum of previous results'
+        description = 'FractionSum of ' + ', '.join([infos[0]['title'] for infos in infoses])
```
Comment (jrising) [inline openest/generate/functions.py:545]: This won't be very interpretable.

File: openest/generate/functions.py [hunk 6]
```diff
@@ -443,6 +443,148 @@ def describe():
...
+        fullinfos = [info for infos in infoses for info in infos]
+        return [dict(name='fractionsum', title=title, description=description)] + fullinfos
+
+    def enable_deltamethod(self):
+        raise NotImplementedError
+
+    def partial_derivative(self, covariate, covarunit):
```
Comment (jrising) [inline openest/generate/functions.py:552]: Any reason you're not implementing this?

File: openest/test/generate/test_functions.py [hunk 7]
```diff
@@ -92,6 +92,85 @@ def test_spaninstabase():
...
+    @pytest.mark.parametrize(
+        "unshift_flag,n_expected",
+        [(True, 3), (False, 1)],
+        ids=['shifted', 'unshifted'],
+    )
+    def test_units_append(self, unshift_flag, n_expected):
+        """Tests that FractionSum correctly appends units from the original subcalcs
+        """
+        unit = ['fakeunit']
+        frac_mock = MockAppCalc(years=[0], values=[0.1], unitses=unit)
```
Comment (jrising) [inline openest/test/generate/test_functions.py:108]: Shouldn't have `fakeunit`.

**Comment by jrising** [comment]
@brews The `describe` is fine, but I also felt like the `apply` was pretty cludgy. May make sense to define a new `Application` subclass that takes a *subcalcs and passes the *result[1]s to a waiting function. That's if you're going to stick to just supporting 3 subcalcs.

File: openest/generate/functions.py (hunk 6 above)
Comment (brews) [reply openest/generate/functions.py:552]: `Product` isn't merged into `master` yet. I'll get on it as soon as it's in.

**Comment by brews** [comment]
Note, I force pushed this after rebasing onto `master` to get the new `Production` class. We need this to implement the partial derivs method.

File: openest/generate/functions.py (hunk 1 above)
Comment (brews) [reply openest/generate/functions.py]: Resolved with acc586a.

File: openest/generate/functions.py (hunk 2 above)
Comment (brews) [reply openest/generate/functions.py:612]: @jrising What are accepted values for unitless units?

File: openest/generate/functions.py (hunk 3 above)
Comment (brews) [reply openest/generate/functions.py:632]: We're actually raising a `NotImplementedError` before we get to this.

I'm not sure how to handle this for the Calculation.

File: openest/generate/functions.py (hunk 4 above)
Comment (brews) [reply openest/generate/functions.py:634]: We're raising a NotImplementedError before we get to this.

File: openest/generate/functions.py (hunk 5 above)
Comment (brews) [reply openest/generate/functions.py:690]: Does this need to be human readable or how is this information used/interpreted? Would you elaborate on how to handle this relatively sophisticated Calculation, in this case?

File: openest/test/generate/test_functions.py (hunk 7 above)
Comment (brews) [reply openest/test/generate/test_functions.py]: What do we use to indicate unitless-ness? Calculations require some form of input for `unitses`, so just a list of `None` (this is how it's now done in 
acc586a).

File: openest/generate/functions.py (hunk 2 above)
Comment (brews) [reply openest/generate/functions.py:613]: Resolved with ec0e39203139fd0632a354e398e6075aea9f8cfe

File: openest/test/generate/test_functions.py (hunk 7 above)
Comment (brews) [reply openest/test/generate/test_functions.py]: Resolved in ec0e39203139fd0632a354e398e6075aea9f8cfe. We're using "unitless".

File: openest/generate/functions.py (hunk 6 above)
Comment (brews) [reply openest/generate/functions.py:697]: Resolved with 20ecad1, should be checked.

File: openest/generate/functions.py (hunk 3 above)
Comment (brews) [reply openest/generate/functions.py]: Resolved with 26b0a1c.

File: openest/generate/functions.py (hunk 4 above)
Comment (brews) [reply openest/generate/functions.py]: Resolved with 26b0a1c.

File: openest/generate/functions.py (hunk 5 above)
Comment (brews) [reply openest/generate/functions.py]: Resolved with 21cc3c0.

**Comment by brews** [comment]
Ready for review, @jrising!

This PR has undergone a heavy rewrite following your feedback.

Please, especially check the new `.partial_derivative()` method - I'm not sure I handle the final weighted value correctly. You might be able to simplify the whole thing.

Thanks for your time. I know you're busy.

**Comment by brews** [comment]
Thanks for the review, @jrising. Appreciate your patience on this.

- > Your 1 - {last_weight} math in format will fail if there are more than 2 terms.

Testing this I get:
```python
>>> subcalc_strs = ['a', 'b', 'c', 'd', 'e']  # b, d are unitless weight subcalcs
>>> last_weight = ' + '.join(subcalc_strs[1:-1:2])
>>> last_weight
'b + d'
>>> last_product = f"(1 - {last_weight}) * {subcalc_strs[-1]}"
>>> last_product
'(1 - b + d) * e'
```
Which looks good to me...? Let me know if I'm missing something obvious. I'll open a new issue and fix this if there is a bug.

- > I change the np.any(weights > 1) check in generate to ensure that sum(weights) < 1.

Changed in deb8e45.

- > Aren't you missing the un-multiplied self.subcalcs[-1] term in your partial derivative calculator?

Ah! Thank you. Fixed in e26227cb.

- > I think the complexity of setting up the expression in order to take its derivative defeats purpose of writing it that way.

I wouldn't fight you if you have any improvements. I mostly wanted to avoid dealing with the parts for the partials of all the products in the calculation.

**Comment by jrising** [comment]
On 1, it should be '(1 - b - d) * e', unless I'm missing something obvious.

**Comment by brews** [comment]
Man, that's sloppy of me. Thank you! Got the bug with PR #38

**Pull request merged**